system/lib/wasmfs/wasmfs.cpp

```diff
@@ -8,60 +8,258 @@
+
+class OpenFile {
+  std::shared_ptr<File> file_node;
+  uint32_t _offset;
+  std::mutex openfile_mutex;
+
+public:
+  OpenFile(uint32_t offset, std::shared_ptr<File> _file_node) {
+    this->_offset = offset;
+    this->file_node = _file_node;
+  }
+
+  std::shared_ptr<File> get_file_node() { return file_node; }
+};
+
+class FileTable {
+  std::vector<std::shared_ptr<OpenFile>> handles;
+  std::mutex ft_mutex;
+
+public:
+  std::vector<std::shared_ptr<OpenFile>> get_file_handles() { return handles; }
+  std::shared_ptr<OpenFile> get_open_file(__wasi_fd_t fd) {
+    // Check if the file descriptor is invalid.
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    if (fd >= handles.size() || fd < 0)
+      return nullptr;
+
+    return handles[fd];
+  }
+
+  __wasi_fd_t add_open_file(std::shared_ptr<OpenFile> ptr) {
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    for (__wasi_fd_t i = 0; i < handles.size(); i++) {
+      if (!handles[i]) {
+        // Free open file entry.
+        handles[i] = ptr;
+        return i;
       }
-      num += len;
     }
-    *nwritten = num;
+
+    // Could not find an empty open file table entry.
+    handles.push_back(ptr);
+
+    return handles.size() - 1;
   }
-  return 0;
+
+  int remove_open_file(__wasi_fd_t fd) {
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    // Check if the file descriptor is invalid.
+    if (fd >= handles.size() || fd < 0)
+      return -1;
+
+    handles[fd] = nullptr;
+    return 0;
+  }
+
+  int set_open_file(__wasi_fd_t fd, std::shared_ptr<OpenFile> ptr) {
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    if (fd < 0)
+      return -1;
+
+    if (fd >= handles.size()) {
+      handles.resize(fd);
+      handles.insert(handles.begin() + fd, ptr);
+    } else {
+      handles[fd] = ptr;
+    }
+
+    return 0;
+  }
+};
+
+// Global filetable singleton
+static FileTable* ft = nullptr;
+
+void populate_file_table() {
+  ft = new FileTable();
+  ft->add_open_file(std::make_shared<OpenFile>(0, std::make_shared<StdinFile>()));
+  ft->add_open_file(std::make_shared<OpenFile>(0, std::make_shared<StdoutFile>()));
+  ft->add_open_file(std::make_shared<OpenFile>(0, std::make_shared<StderrFile>()));
+}
+
+__wasi_fd_t __syscall63(__wasi_fd_t oldfd, __wasi_fd_t newfd) { // dup2
+  if (newfd < 0) {
+    return -1;
+  }
+  if (!ft) {
+    populate_file_table();
+  }
+
+  if (oldfd == newfd) {
+    return oldfd;
+  }
+
+  // If oldfd is not a valid file descriptor, then the call fails,
+  // and newfd is not closed.
+
+  std::shared_ptr<OpenFile> oldfd_openfile = ft->get_open_file(oldfd);
+  // Return -1 if the fd is invalid or points to a nullptr
+  if (oldfd_openfile == nullptr)
+    return -1;
+
+  // If the file descriptor newfd was previously open, it is closed
+  // before being reused; the close is performed silently.
+
+  std::shared_ptr<OpenFile> newfd_openfile = ft->get_open_file(newfd);
+  if (newfd_openfile) {
+    ft->remove_open_file(newfd);
+  }
```

Comment:
Closing and reusing `newfd` needs to happen atomically, so the file table lock needs to be held over both operations. Probably the cleanest/simplest way to do that is to make the caller of all `FileTable` methods responsible for locking rather than having the methods themselves do the locking.

Reply:
If you want to get fancy and make it impossible to perform file table operations without holding its lock, you can force all operations to go through an RAII wrapper object that holds the lock for the duration of its lifetime. The method for getting that wrapper object could also be responsible for lazily initializing the table.

Here's a sketch:

```C++
class FileTable {
  std::mutex mutex;
  std::vector<...> entries;
  FileTable() { /* initialize std streams */ }
public:
  class Handle {
    FileTable& file_table;
    std::lock_guard<std::mutex> lock;
  public:
    Handle(FileTable& file_table) : file_table(file_table), lock(file_table.mutex) {} 
    // ... file table accessors that internally operate on file_table.entries ...
  };
  Handle get();
};
FileTable::Handle FileTable::get() {
  static FileTable file_table;
  return FileTable::Handle{file_table, file_table.mutex};
}
```

This has the following properties:

1. The `FileTable` constructor is private, so no other code can create a `FileTable`.
2. The `file_table` member of `Handle` is private, so no other code can get a reference to a `FileTable`.
3. Due to (1) and (2),  all accesses to a `FileTable` must go through a `Handle`.
4. A `Handle` hold its `FileTable`'s lock for the duration of its lifetime.
5. Due to (3) and (4), _all accesses to a `FileTable` happen while holding its lock._
6. A reference to a `FileTable` is necessary to construct a `Handle`.
7. Due to (1), (2), and (6) the only way to get a `Handle` is by calling `FileTable::get()`.
8. `FileTable::get()` always returns a handle to the same lazily initialized `FileTable`.
9. Due to (3), (7), and (8), _there is only ever one `FileTable` in the system and it is lazily initialized._

For even more fun, you can make the API for accessing the `FileTable` super ergonomic by implementing `FileTable::Handle::operator[]()`.